Implement a function that gives access to ID of currently executed task

b-naber opened the pull request:
Fixes https://github.com/tokio-rs/tokio/issues/1996

This PR introduces a `tokio-unstable` function that allows one to get the `ID` of the currently executed task.

This introduces a new thread local `TASK_ID`, which is set before tasks are executed, to track which `Id` is currently executed on a worker. Had to move `task_id` from `Core` to `Header` in order to allow access to the id in a `RawTask`.

hawkw commented on tokio/tests/task_local.rs:
```diff
@@ -31,3 +31,16 @@ async fn local() {
     j3.await.unwrap();
 }
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn test_task_id() {
+    use tokio::task;
+
+    let handle = tokio::spawn(async {
+        let id = task::current_id();
+        println!("current task id: {}", id);
+    });
+
+    handle.await.unwrap();
+}
```
we should probably have some tests that two or more tasks don't get colliding IDs.

hawkw replied:
also, we could probably test that `current_id()` and `handle.id()` return the same value if we gave the spawned task a oneshot channel, like this:
```suggestion
}

#[cfg(tokio_unstable)]
#[tokio::test(flavor = "current_thread")]
async fn task_ids_match() {
    use tokio::{task, sync::oneshot};
    
    let (tx, rx) = oneshot::channel();
    let handle = tokio::spawn(async {
        let id = rx.await.unwrap();
        assert_eq!(id, task::current_id());
    });

    tx.send(handle.id());

    handle.await.unwrap();
}
```

hawkw commented on tokio/src/runtime/context.rs:
```diff
@@ -1,12 +1,15 @@
     static CONTEXT: RefCell<Option<Handle>> = RefCell::new(None)
 }
 
+thread_local! {pub(crate) static TASK_ID: Cell<Option<Id>> = Cell::new(None) }
```
it would be preferable to use an existing thread local value to store the current task ID, if possible, because some systems have a limit on the number of thread local variables that may be created. we may want to change the existing `CONTEXT` thread-local into a `struct`, like this:

```rust
thread_local! {
    static CONTEXT: Context = Context {
        handle: RefCell::new(None),
        current_id: Cell::new(None), 
    };
}

struct Context {
    handle: RefCell<Option<Handle>>,
    current_id: Cell<Option<Id>>,
}
```
this way, we would only use a single thread-local key, rather than two.

hawkw commented on tokio/src/runtime/basic_scheduler.rs:
```diff
@@ -536,12 +536,17 @@ impl CoreGuard<'_> {
                         task.run();
                     });
 
+                    context::set_current_task_id(old_id);
```
we may want to do this via a `Drop` guard, rather than with a pair of function calls, to ensure that the previous task ID is always reset, even in the event of a panic in Tokio internals...

hawkw commented on tokio/src/runtime/task/harness.rs:
```diff
@@ -116,14 +123,16 @@ where
                         // cancelled during the poll.
                         let core = self.core();
-                        cancel_task(&core.stage, core.task_id.clone());
+                        let task_id = self.get_task_id();
```
can we reuse the existing `task_id` variable here?

hawkw reviewed (commented):
some minor notes, hope they're helpful!

b-naber commented:
@hawkw Thanks for the review. Addressed your comments.

I can't reproduce the test failure of `test_task_id_colliding` locally. 

There is a size assertion in one of the tests for `Header` which fails on 32-bit systems. Not sure what's the right fix here, probably a conditional type choice for the inner type of `Id`(u32 on 32-bit systems)?

How can I make the doc test conditional on `tokio-unstable` (took it out for now)?

b-naber commented:
@hawkw The previous `Header` layout with `tokio-unstable` and `tracing` enabled also doesn't fit into the cache line on 32-bit systems. 

Can you comment on how important it is to be able to do that?

hawkw commented:
> @hawkw The previous `Header` layout with `tokio-unstable` and `tracing` enabled also doesn't fit into the cache line on 32-bit systems.
> 
> Can you comment on how important it is to be able to do that?

It's fairly important. Perhaps we should consider moving one of those fields out of the header --- since the `tracing` ID is only used when actually polling the task (and not dereferenced through other task pointers such as `JoinHandle`), we might want to consider moving it out of the `Header` and into the `Core`.

TennyZhuang commented on tokio/src/runtime/task/mod.rs:
```diff
@@ -503,7 +514,7 @@ impl Id {
             use crate::loom::sync::Mutex;
 
-            static NEXT_ID: Lazy<Mutex<u64>> = Lazy::new(|| Mutex::new(1));
+            static NEXT_ID: Lazy<Mutex<u64>> = Lazy::new(|| Mutex::new(1) );
```
The change is a typo?

Darksonn commented:
Since this PR appears to have gotten stuck, I'll close it for now. Feel free to reopen if you find a solution to the header being too large.

b-naber commented:
@Darksonn @hawkw Since it's been almost a month since this has been marked 'ready for review' and maybe that was missed by you, just pinging to say that this is ready and can be reviewed.

Darksonn commented on tokio/src/runtime/task/core.rs:
```diff
@@ -50,8 +50,9 @@ pub(super) struct Core<T: Future, S> {
 
-    /// The task's ID, used for populating `JoinError`s.
-    pub(super) task_id: Id,
+    /// The tracing ID for this instrumented task.
+    #[cfg(all(tokio_unstable, feature = "tracing"))]
+    pub(super) id: Option<tracing::Id>,
```
I'm a bit confused about what happened to the two id fields here. Were they swapped (one in header other in core)?

b-naber replied:
Yes, originally this was done to potentially allow the `Header` size to fit into a cache line, though even the swap didn't fulfill that on 32-bit system. Now that the `owned` field was moved out of `Header` I believe there should be enough space to move the tracing id field back to `Header`, though since this field is only accessed in contexts in which we _can_ cast a `Header` back into `Core`, it probably makes sense to keep it there?!

b-naber replied:
Actually, to clarify this was done since we need `task_id` in contexts in which we cannot cast `Header` back to `Cell` (at least without fairly large changes), so `task_id` has to be in `Header`. We moved `id` out of `Header` as an attempt to solve the cache line size problem.

Darksonn replied:
I guess it's fine.

Darksonn reviewed (commented):
It mostly looks fine. There are some rustfmt issues.

Darksonn reviewed (approved):
LGTM, but the merge conflicts need to be fixed.

Darksonn commented on tokio/src/task/task_local.rs:
```diff
@@ -446,3 +447,17 @@ impl From<std::thread::AccessError> for ScopeInnerErr {
     }
 }
+
+/// Returns the `Id` of the currently executing task.
+///
+/// # Panics
+///
+/// This function panics if called from outside a task context
+///
+#[cfg_attr(docsrs, doc(cfg(all(feature = "rt", tokio_unstable))))]
+#[cfg_attr(not(tokio_unstable), allow(unreachable_pub))]
+#[allow(dead_code)]
+pub fn current_id() -> Id {
+    use crate::runtime::context;
+    context::current_task_id()
+}
```
Why is this in the `task_local.rs` file? It has nothing to do with task locals as far as I can tell.

Darksonn reviewed (commented):
The extra whitespace that @TennyZhuang pointed out, as well as the extra lines in `src/task/mod.rs` should be fixed. Once that happens, I would be happy to merge this.

Darksonn commented on tokio/src/runtime/task/mod.rs:
```diff
@@ -201,10 +202,23 @@ use std::{fmt, mem};
+/// This function panics if called from outside a task context
+///
+#[cfg_attr(docsrs, doc(cfg(all(feature = "rt", tokio_unstable))))]
+#[cfg_attr(not(tokio_unstable), allow(unreachable_pub))]
+#[allow(dead_code)]
+pub fn current_id() -> Id {
```
Why is there an `#[allow(dead_code)]`?

Darksonn reviewed (commented):
Sorry to keep posting comments. Every time I look over the PR, I see some new super minor thing that I missed the previous time for some reason.

Darksonn commented on tokio/src/runtime/context.rs:
```diff
@@ -90,7 +115,7 @@ pub(crate) fn enter(new: Handle) -> EnterGuard {
 pub(crate) fn try_enter(new: Handle) -> Option<EnterGuard> {
     let rng_seed = new.inner.seed_generator().next_seed();
-    let old_handle = CONTEXT.try_with(|ctx| ctx.borrow_mut().replace(new)).ok()?;
+    let old_handle = CONTEXT.try_with(|ctx| ctx.handle.borrow_mut().replace(new)).ok()?;
```
Apply rustfmt:
```suggestion
    let old_handle = CONTEXT
        .try_with(|ctx| ctx.handle.borrow_mut().replace(new))
        .ok()?;
```

b-naber commented:
Thanks. Is this mergeable now?

vitruvvius commented:
...changes are approved

Darksonn commented on tokio/src/runtime/context.rs:
```diff
@@ -24,11 +33,27 @@ pub(crate) fn current() -> Handle {
+pub(crate) fn current_task_id() -> Id {
+    match CONTEXT.try_with(|ctxt| ctxt.task_id.get()) {
+        Ok(Some(id)) => id,
+        Ok(None) => panic!("tried to get current task id from outside of task"),
+        Err(e) => panic!(
+            "tried to get current task id from outside of task failed with error {:?}",
+            e
+        ),
```
There isn't any reason to print the thread-local error here. The reason is the same either way - we are not inside the task.
```suggestion
        _ => panic!("tried to get current task id from outside of task"),
```

Darksonn commented on tokio/src/runtime/scheduler/current_thread.rs:
```diff
@@ -569,6 +569,8 @@ impl CoreGuard<'_> {
                         }
                     };
 
+                    let task_id = task.get_task_id();
+                    let _task_id_guard = TaskIdGuard::new(task_id);
                     let task = context.handle.shared.owned.assert_owner(task);
```
Why is this being set in the current-thread runtime. It should happen somewhere in the harness code instead.

b-naber replied:
Why should it be set in the harness code? Don't know the harness code that well, but doesn't this just implement the polling and waking functionality of tasks?

Seems fairly reasonable to me to set the task id before running a task.

Darksonn replied:
It's important that we always remember to set the task id, and we shouldn't do it more than once. Setting it in the harness code, which is also responsible for e.g. catching panics when polling the task, makes sure that we have to do it only in one place while making sure that we can't forget.

For example, in this PR, you set it in the current-thread runtime, but _not_ in the multi-thread runtime. However, you also set it in the `run` function, which is called for both current-thread and multi-thread, so you're actually setting it twice for current-thread.

b-naber replied:
Makes sense, I'll try to change this later today.

Darksonn commented on tokio/src/runtime/task/mod.rs:
```diff
@@ -381,6 +403,7 @@ impl<S: Schedule> LocalNotified<S> {
     /// Runs the task.
     pub(crate) fn run(self) {
         let raw = self.task.raw;
+        let _guard = TaskIdGuard::new(raw.get_task_id());
```
This also seems like the wrong place to put this.

Darksonn commented on tokio/src/runtime/task/mod.rs:
```diff
@@ -418,6 +441,8 @@ impl<S: Schedule> UnownedTask<S> {
             _p: PhantomData,
         };
 
+        let _guard = TaskIdGuard::new(task.get_id());
```
Same here.

b-naber commented:
Dont have the time to work on this anymore. Feel free to merge as is.
